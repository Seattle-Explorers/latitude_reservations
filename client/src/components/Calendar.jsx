import React from 'react';
import axios from 'axios';
import moment from 'moment';
import { FaRegKeyboard } from 'react-icons/fa';

import LeftCalendar from './LeftCalendar.jsx';
import RightCalendar from './RightCalendar.jsx';

class Calendar extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      availDates: [],
      currentDate: moment(new Date()),
      firstDate: '',
      secondDate: '',
      nextDate: false,
    };
    this.calendarClickHandler = this.calendarClickHandler.bind(this);
    this.clearButtonClickHandler = this.clearButtonClickHandler.bind(this);
    this.handleRightArrowClick = this.handleRightArrowClick.bind(this);
    this.handleLeftArrowClick = this.handleLeftArrowClick.bind(this);
  }

  componentDidMount() {
    this.fetchDatesForSelectedListingID();
  }

  fetchDatesForSelectedListingID() {
    const { listingID } = this.props;
    return axios.get(`/api/reservation/${listingID}`)
      .then(({ data }) => {
        console.log('GET Request Successful: ', data);
        this.setState({
          availDates: data.open_dates,
        });
      })
      .catch((err) => {
        console.log('Error Fetching Data: ', err);
      });
  }

  calendarClickHandler(targetMonth, targetDate) {
    const thisYear = moment(new Date()).format('YYYY');
    if (targetDate < 10) {
      targetDate = `0${targetDate}`;
    }
    this.setState((prevState) => ({
      nextDate: !prevState.nextDate,
    }));
    if (!this.state.nextDate) {
      this.setState({
        firstDate: `${thisYear}-${targetMonth}-${targetDate}`,
      });
    } else {
      this.setState({
        secondDate: `${thisYear}-${targetMonth}-${targetDate}`,
      });
    }
    console.log('firstDate: ', this.state.firstDate, 'secondDate: ', this.state.secondDate);
  }

  clearButtonClickHandler() {
    this.setState({
      firstDate: '',
      secondDate: '',
    });
  }

  handleRightArrowClick() {
    this.setState({
      currentDate: moment(this.state.currentDate).add(1, 'months'),
    });
  }

  handleLeftArrowClick() {
    this.setState({
      currentDate: this.state.currentDate.subtract(1, 'months'),
    });
  }

  render() {
    return (
      <div className="calendar-main-container">
        <div className="calendar-top-descrpiton">
          <h2>Select Check-in date</h2>
          <div>
            This host offers 10% off if you stay a week and a 20% monthly discount.
          </div>
        </div>
        <div className="calendars">
          <LeftCalendar
            availDates={this.state.availDates}
            currentDate={this.state.currentDate}
            handleClick={this.calendarClickHandler}
            handleLeftArrowClick={this.handleLeftArrowClick}
            handleRightArrowClick={this.handleRightArrowClick}
            clearButtonClickHandler={this.clearButtonClickHandler}
          />
          <RightCalendar
            availDates={this.state.availDates}
            currentDate={this.state.currentDate}
            handleClick={this.calendarClickHandler}
            handleArrowClick={this.handleRightArrowClick}
          />
        </div>
        <div className="calendar-bottom">
          <div className="keyboard-icon">
            <FaRegKeyboard size={32} />
          </div>
          <div className="clear-dates-btn-container">
            <div className="clear-dates-space" />
            <button className="clear-dates-btn" onClick={this.clearButtonClickHandler}>Clear dates</button>
          </div>
        </div>
      </div>
    );
  }
}

export default Calendar;
