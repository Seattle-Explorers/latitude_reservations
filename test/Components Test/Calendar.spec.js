import React from 'react';
import { shallow, configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import Calendar from '../../client/src/components/Calendar.jsx';

require('babel-polyfill');

configure({ adapter: new Adapter() });

describe('Calendar Component', () => {
  jest.setTimeout(90000);
  it('should render correctly when given a valid listing ID', async () => {
    const component = await shallow(<Calendar listingID="025" />);

    expect(component).toMatchSnapshot();
  });
});
