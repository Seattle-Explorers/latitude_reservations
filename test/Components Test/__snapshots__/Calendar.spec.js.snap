// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Calendar Component should render correctly when given a valid listing ID 1`] = `ShallowWrapper {}`;

exports[`Calendar Component should render correctly with no props 1`] = `ShallowWrapper {}`;
